### DISCUSSION TITLE
Need Help Debugging "Stack too deep" Error in Solidity with Forge Coverage

### DISCUSSION THREAD

**Discussion by danstam** [Cyfrin/foundry-full-course-cu#851] [Q&A] [answered]
Solidity  Version: 0.8.18

I'm currently working through the course and I've run into a bit of a roadblock. When I run `forge coverage` on my project, I'm getting a "Stack too deep" error. I am aware that this error typically occurs when a function uses more than 16 local variables, which is a limitation of the EVM.

The problem is, the error message doesn't specify where the issue is coming from, making it difficult for me to pinpoint the problematic function. I've tried searching through my codebase for complex functions or ones with many local variables, but I haven't had any luck so far.I'm starting to wonder if the error could be coming from a file that is part of one of the imported libraries. 

I also tried adding `viaIR: true` to my `foundry.toml` file. This allows the compiler to optimize the code more effectively, potentially reducing the number of local variables that get put on the stack but this isn't working either.

Interestingly, when I run `forge test`, everything works fine and no errors are thrown. This issue only arises when I run `forge coverage`.

Has anyone encountered a similar issue before or have any suggestions on how I might debug this? Any help would be greatly appreciated!

**Comment by 0xusmanf** [comment]
Can you give us your GitHub repo link so we can recreate the issue?

**Reply by danstam** [reply]
[This](https://github.com/danstam/foundry-smart-contract-lottery) is the project that is giving me the issue

**Reply by 0xusmanf** [reply]
I think you have pushed the wrong project to this repo. Please check and provide link with the correct project.

**Reply by danstam** [reply]
[This](https://github.com/danstam/foundry-smart-contract-lottery) should open the repo.

**Reply by 0xusmanf** [reply, marked as answer]
**Stack too deep** error occurs due to the following reasons.

1. The EVM stack has 1024 slots available and every time a function calls another function, it reserves some space on the stack to store arguments, return values, and local variables. The deeper the call chain, the more slots you need and you eventually run out of slots if you go too deep.

1. OR you try to read from deeper than the 16th slot.

In your case 2nd seems to be the reason because if you check `(/solidity/libsolidity/codegen/CompilerUtils.cpp:1430)` the following `assertThrow` is triggering.

https://github.com/ethereum/solidity/blob/a1b79de64235f13e6b06e088fe6365c5a12d13d3/libsolidity/codegen/CompilerUtils.cpp#L1428-L1437

Trying to receive too many variables is causing this error in the following `setUp()` function of `RaffleTest.t.sol`.

https://github.com/danstam/foundry-smart-contract-lottery/blob/2202cbae4a62ad32decc14cb3701e37fd409f022/test/unit/RaffleTest.t.sol#L38-L53

You can simply get around this by removing one of the variables you may not need like I have removed `deployerKey` in code below.
```solidity
        (
            subscriptionId,
            gasLane,
            automationUpdateInterval,
            raffleEntranceFee,
            callbackGasLimit,
            vrfCoordinatorV2,
            link,
        ) = helperConfig.activeNetworkConfig();
```
Or if you need all of the variables you can also divide it into 2 parts like below.
```solidity
        (
            subscriptionId,
            gasLane,
            automationUpdateInterval,
            raffleEntranceFee,
            callbackGasLimit,
            vrfCoordinatorV2,
            link,
        ) = helperConfig.activeNetworkConfig();
        (,,,,,,, deployerKey) = helperConfig.activeNetworkConfig();
```

After implementing this you don't have to use `viaIR: true`, because it takes ages to compile.

**Reply by danstam** [reply]
Thank you immensely!

**Reply by theGuy-03** [reply]
thanks Usman Farooq really helped alot

**Reply by SebasAran16** [reply]
Thank you a lot! Was getting crazy about finding where the issue was. Now is much clearer the problem too!
